atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultBucketCloseTimestampCalculatorTest.java

```diff
@@ -0,0 +1,163 @@
+
+import static com.palantir.atlasdb.sweep.asts.bucketingthings.DefaultBucketCloseTimestampCalculator.MAX_BUCKET_SIZE_FOR_NON_PUNCHER_CLOSE;
+import static com.palantir.atlasdb.sweep.asts.bucketingthings.DefaultBucketCloseTimestampCalculator.MIN_BUCKET_SIZE_FOR_NON_PUNCHER_CLOSE;
+import static com.palantir.atlasdb.sweep.asts.bucketingthings.DefaultBucketCloseTimestampCalculator.TIME_GAP_BETWEEN_BUCKET_START_AND_END;
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+import static org.mockito.Mockito.when;
+
+import com.palantir.atlasdb.cleaner.PuncherStore;
+import com.palantir.atlasdb.sweep.asts.bucketingthings.DefaultBucketCloseTimestampCalculator.CloseableState;
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+public final class DefaultBucketCloseTimestampCalculatorTest {
+    private final AtomicLong freshTimestamp = new AtomicLong(0);
+    private final FakeClock clock = new FakeClock();
+
+    @Mock
+    private PuncherStore puncherStore;
+
+    private DefaultBucketCloseTimestampCalculator bucketCloseTimestampCalculator;
+
+    @BeforeEach
+    public void setup() {
+        bucketCloseTimestampCalculator =
+                new DefaultBucketCloseTimestampCalculator(puncherStore, freshTimestamp::get, clock);
+    }
+
+    @Test
+    public void returnsNotCloseableIfSufficientTimeHasNotPassedSinceStartTimestamp() {
+        int startTimestamp = 18; // Arbitrarily chosen.
+        when(puncherStore.getMillisForTimestamp(startTimestamp)).thenReturn(clock.millis());
+        CloseableState state = bucketCloseTimestampCalculator.getCloseableStateForStartTimestamp(startTimestamp);
+        assertThat(state.canBeClosed()).isFalse();
+    }
+
+    @Test
+    public void
+            returnsLogicalTimestampSufficientTimeAfterStartTimestampIfTenMinutesHasPassedAndLogicalTimestampAheadOfStart() {
+        int startTimestamp = 123;
+        when(puncherStore.getMillisForTimestamp(startTimestamp)).thenReturn(clock.millis());
+        clock.advance(TIME_GAP_BETWEEN_BUCKET_START_AND_END);
+        when(puncherStore.get(clock.millis())).thenReturn(153L);
+
+        CloseableState state = bucketCloseTimestampCalculator.getCloseableStateForStartTimestamp(startTimestamp);
+        assertThat(state.canBeClosed()).isTrue();
+        assertThat(state.endTimestampExclusive()).isEqualTo(153L);
+    }
+
+    @ParameterizedTest
+    @ValueSource(longs = {2300, 2315}) // less than, and equal to.
```

Comment:
discussed offline: not sure I follow, though this is used to test the situation where the clock "drift" means that our naive usage of the puncher wants to create an [n, n) range